torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -1389,13 +1403,96 @@ void ONNXShapeTypeInference(
 
 } // namespace
 
+std::pair<bool, bool> IsShapeReliable(Node* n) {
+  auto reliable = true;
+  auto complete = true;
+  for (auto input : n->inputs()) {
+    if (ConstantValueMap::HasTypeReliable(input->debugName())) {
+      auto reliable_value =
+          ConstantValueMap::GetTypeReliable(input->debugName()).value();
+      reliable = reliable && reliable_value;
+    } else {
+      reliable = false;
+    }
+    if (auto pt = input->type()->cast<TensorType>()) {
+      if (!pt->sizes().isComplete()) {
+        complete = false;
+      }
+    }
+  }
+  return std::make_pair(reliable, complete);
+}
+
+// There is no need to put onnx type here, but we need this
+// for some legacy tests when onnx_shape_inference=False.
+static std::unordered_set<std::string> nodeTypeReliableForTracer = {
+    "prim::ListConstruct",
+    "onnx::Cast",
+    "onnx::Constant",
+    "com.microsoft::Gelu"};
+
+void UpdateReliable(
+    torch::jit::Value* output,
+    const std::pair<bool, bool>& inferred_type_reliable) {
+  auto inferred = false;
+  if (ConstantValueMap::HasUseInferredType(output->debugName())) {
+    inferred =
+        ConstantValueMap::GetUseInferredType(output->debugName()).value();
+  }
+  auto isTypeReliableForTracer =
+      nodeTypeReliableForTracer.find(
+          output->node()->kind().toDisplayString()) !=
+      nodeTypeReliableForTracer.end();
+  if (!inferred && !isTypeReliableForTracer &&
+      !output->node()->kind().is_onnx()) {
+    std::cerr
+        << "WARNING: The shape inference of "
+        << output->node()->kind().toDisplayString()
+        << " type is missing, so it may result in wrong shape inference for the exported graph. "
+        << "Please consider adding it in symbolic function." << std::endl;
+  }
+  auto reliable = false;
+  if (inferred) {
```

Comment:
can you explain the logic here, and what does the two bool values in `inferred_type_reliable`  represent?

Reply:
The tensor is estimated by inference if `inferred` is true, by tracer if it is false.
`inferred_type_reliable.first` is true means all the node input tensors are reliable.
`inferred_type_reliable.second` is true means all the node input tensors are static.
This piece of code is to determine whether the current node output tensor is reliable, and update the reliable map.
Using the following logic:
If it is estimated by inference, it is reliable when `inferred_type_reliable.first` is true, vice versa.
If it is estimated by tracer, it is reliable when `inferred_type_reliable.second` is true && this node type is in the nodeTypeReliableForTracer set.
In addition,   
// Assume that the tracer can estimate rank correctly,
// then the output tensor of Shape should always be reliable.